File: day8/day8.py
#!/usr/bin/env python3

letters = "abcdefg"
segments = [
    "abcefg",
    "cf",
    "acdeg",
    "acdfg",
    "bcdf",
    "abdfg",
    "abdefg",
    "acf",
    "abcdefg",
    "abcdfg",
]

def encode(lseg):
    nseg = 0
    for l in lseg:
        nseg |= 1 << letters.index(l)
    return nseg

def count_ones(i):
    return bin(i).count("1")

# common segments between digits
intersections = [[count_ones(encode(segments[i]) & encode(segments[j]))
                  for i in range(10)]
                 for j in range(10)]

# map a segment length to candidates numbers
candidates = {}
for i, seg in enumerate(segments):
    l = len(seg)
    candidates[l] = candidates.get(l, []) + [i]

def part1(values):
    count = 0
    for _, display in values:
        for value in display:
            if len(candidates[count_ones(value)]) == 1:
                count += 1
    return count

def solve(row):
    wiring, display = row

    # find the candidates for each digit
    nums = {}
    for wire in wiring:
        for i in candidates[count_ones(wire)]:
            nums[i] = nums.get(i, []) + [wire]

    # solve the constraints
    order = sorted(range(10), key = lambda x: len(nums[x]))
    t = 0
    while len(nums[order[t]])==1:
        t += 1

    k = 0
    while k < t:
        i = order[k]
        iseg = nums[i][0]
        for l in range(t, 10):
            j = order[l]
            nums[j] = [jseg for jseg in nums[j] if count_ones(iseg & jseg) == intersections[i][j]]
            if len(nums[j]) == 1:
                order[t], order[l] = order[l], order[t]
                t += 1
        k += 1

    if t < 10:
        print("Couldn't find a solution")
        return 0

    # map the encoded digit to the the actual value
    segmap = {v[0]: k for k, v in nums.items()}

    # compute the value in the display
    value = 0
    for digit in display:
        value = value * 10 + segmap[digit]

    return value

def part2(values):
    return sum(solve(x) for x in values)

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        values = []
        for row in txt.strip().splitlines():
            wiring, display = row.split(" | ")
            values.append(
                ([encode(x) for x in wiring.split()],
                 [encode(x) for x in display.split()])
            )

        print("Part1:", part1(values))
        print("Part2:", part2(values))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)

File: day8/day8.cpp
#include <algorithm>
#include <numeric>
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

// entry in the input
struct Entry
{
	unsigned wiring[10];
	unsigned display[4];
};

// for a given number of segments turned on
// provide a list of potential digits
static vector<unsigned> candidates[8];

// number of common segments for two digits
static int inters[10][10];

static unsigned encode(const string& str)
{
	unsigned value = 0;
	for (auto s: str)
	{
		value |= 1U<<(s-'a');
	}
	return value;
}

istream& operator>>(istream& in, Entry& e)
{
	string tmp;
	for (int i = 0; i < 10; ++i)
	{
		in >> tmp;
		e.wiring[i] = encode(tmp);
	}
	in >> tmp;		// |
	for (int i = 0; i < 4; i++)
	{
		in >> tmp;
		e.display[i] = encode(tmp);
	}
	return in;
}

static int bitcount(unsigned x)
{
	int count = 0;
	while (x)
	{
		x &= x - 1;
		count++;
	}
	return count;
}

static unsigned part1(const vector<Entry>& entries)
{
	unsigned count = 0;
	for (const auto& e : entries)
	{
		for (const auto &d: e.display)
		{
			if (candidates[bitcount(d)].size() == 1)
			{
				count++;
			}
		}
	}
	return count;
}

static unsigned solve(const Entry& entry)
{
	// find each digit find the list of wiring candidates
	vector<unsigned> digits[10];
	for (auto& w: entry.wiring)
	{
		int bc = bitcount(w);
		for (auto i: candidates[bc])
		{
			digits[i].push_back(w);
		}
	}

	// order the digits by the number of candidates
	size_t ordered[10];
	for (size_t i = 0; i < 10; i++)
	{
		ordered[i] = i;
	}
	sort(ordered, ordered + 10, [&digits](int i, int j)
	{
		return digits[i].size() < digits[j].size();
	});

	// find the index where the candidates > 1
	size_t t = 0;
	while (digits[ordered[t]].size() <= 1)
	{
		t++;
	}

	// remove the wrong candidates
	for (size_t i = 0; i < t; i++)
	{
		unsigned a = ordered[i];
		unsigned aseg = digits[a][0];

		for (size_t j = t; j < 10; j++)
		{
			unsigned b = ordered[j];
			for (auto it = digits[b].begin(); it != digits[b].end();)
			{
				int inter = bitcount(aseg & *it);
				if (inter != inters[a][b])
				{
					it = digits[b].erase(it);
				}
				else
				{
					++it;
				}
			}
			if (digits[b].size() == 1)
			{
				swap(ordered[j], ordered[t]);
				t++;
			}
		}
	}

	if (t < 10)
	{
		fmt::print(stderr, "Couldn't find a solution\n");
		return 0;
	}

	// map the encoded segment value to a digit
	unsigned map[1U<<8];
	for (size_t i = 0; i < 10; i++)
	{
		map[digits[ordered[i]][0]] = ordered[i];
	}

	// compute the result
	return accumulate(
		entry.display, entry.display+4, 0,
		[map](size_t acc, unsigned e)
		{
			return acc * 10 + map[e];
		});
}

static size_t part2(const vector<Entry>& entries)
{
	return accumulate(entries.cbegin(),
			  entries.cend(),
			  0,
			  [](size_t acc, const Entry& e)
			  {
				  return acc + solve(e);
			  });
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Entry> entries;
	Entry e;
	while (in>>e)
	{
		entries.push_back(move(e));
	}
	in.close();

	static const string segs[10] = {
		"abcefg",
		"cf",
		"acdeg",
		"acdfg",
		"bcdf",
		"abdfg",
		"abdefg",
		"acf",
		"abcdefg",
		"abcdfg",
	};

	// prepare the data
	for (size_t i = 0; i < 10; ++i)
	{
		candidates[segs[i].size()].push_back(i);
		for (size_t j = 0; j < 10; ++j)
		{
			inters[i][j] = bitcount(
				encode(segs[i]) & encode(segs[j]));
		}
	}

	fmt::print("Part1: {}\n", part1(entries));
	fmt::print("Part2: {}\n", part2(entries));
	return 0;
}
